packages/firestore/test/unit/local/query_engine.test.ts

```diff
@@ -773,6 +773,466 @@ function genericQueryEngineTest(
+    }
+
+    const query1 = query(
+      'coll',
+      orFilter(filter('a', 'in', [2, 3]), filter('b', 'array-contains', 3))
+    );
+    const result1 = await expectFunction(() =>
+      runQuery(query1, lastLimboFreeSnapshot)
+    );
+    verifyResult(result1, [doc3, doc4, doc6]);
+
+    const query2 = query(
+      'coll',
+      andFilter(filter('a', 'in', [2, 3]), filter('b', 'array-contains', 7))
+    );
+    const result2 = await expectFunction(() =>
+      runQuery(query2, lastLimboFreeSnapshot)
+    );
+    verifyResult(result2, [doc3]);
+
+    const query3 = query(
+      'coll',
+      orFilter(
+        filter('a', 'in', [2, 3]),
+        andFilter(filter('b', 'array-contains', 3), filter('a', '==', 1))
+      )
+    );
+    const result3 = await expectFunction(() =>
+      runQuery(query3, lastLimboFreeSnapshot)
+    );
+    verifyResult(result3, [doc3, doc4, doc6]);
+
+    const query4 = query(
+      'coll',
+      andFilter(
+        filter('a', 'in', [2, 3]),
+        orFilter(filter('b', 'array-contains', 7), filter('a', '==', 1))
+      )
+    );
+    const result4 = await expectFunction(() =>
+      runQuery(query4, lastLimboFreeSnapshot)
+    );
+    verifyResult(result4, [doc3]);
+  });
+
+  it('order by equality', async () => {
+    const doc1 = doc('coll/1', 1, { 'a': 1, 'b': [0] });
+    const doc2 = doc('coll/2', 1, { 'b': [1] });
+    const doc3 = doc('coll/3', 1, { 'a': 3, 'b': [2, 7], 'c': 10 });
+    const doc4 = doc('coll/4', 1, { 'a': 1, 'b': [3, 7] });
+    const doc5 = doc('coll/5', 1, { 'a': 1 });
+    const doc6 = doc('coll/6', 1, { 'a': 2, 'c': 20 });
+    await addDocument(doc1, doc2, doc3, doc4, doc5, doc6);
+
+    let expectFunction = expectFullCollectionQuery;
+    let lastLimboFreeSnapshot = MISSING_LAST_LIMBO_FREE_SNAPSHOT;
+
+    if (durable) {
+      expectFunction = expectOptimizedCollectionQuery;
+      lastLimboFreeSnapshot = SnapshotVersion.min();
+
+      await indexManager.addFieldIndex(
+        fieldIndex('coll', { fields: [['a', IndexKind.ASCENDING]] })
+      );
+      await indexManager.addFieldIndex(
+        fieldIndex('coll', { fields: [['a', IndexKind.DESCENDING]] })
+      );
+      await indexManager.addFieldIndex(
+        fieldIndex('coll', { fields: [['b', IndexKind.CONTAINS]] })
+      );
+      await indexManager.updateIndexEntries(
+        documentMap(doc1, doc2, doc3, doc4, doc5, doc6)
+      );
+      await indexManager.updateCollectionGroup(
+        'coll',
+        newIndexOffsetFromDocument(doc6)
+      );
+    }
+
+    const query1 = query('coll', filter('a', '==', 1), orderBy('a'));
+    const result1 = await expectFunction(() =>
+      runQuery(query1, lastLimboFreeSnapshot)
+    );
+    verifyResult(result1, [doc1, doc4, doc5]);
+
+    const query2 = query('coll', filter('a', 'in', [2, 3]), orderBy('a'));
+
+    const result2 = await expectFunction(() =>
+      runQuery(query2, lastLimboFreeSnapshot)
+    );
+    verifyResult(result2, [doc6, doc3]);
+  });
+
```

Comment:
Thanks for adding all the additional tests.

IIUC `durable` means the test is using memory persistence.

so, what we're testing here are:
- queries using full collection scan (using memory persistence)
- queries using client-side indexing (using IndexedDB)

But, what we really need to have is:
- queries using full collection scan (using memory persistence)
- queries using full collection scan (using IndexedDB)
- queries using client-side indexing (using IndexedDB)

For Android, this is the case.
- Tests in QueryEngineTestCase.java are run twice, once for memory persistence (full collection scan + memory persistence) and once with SQLite persistence (full collection scan + sqlite).
- Tests in SQLiteQueryEngineTest.java are run once (CSI + sqlite)

Reply:
I refactored query_engine.test.ts so the tests execute in three configurations: 
1. durable with client side indexing
2. durable without client side indexing
3. non-durable without client side indexing

PTAL. Sorry, it's a little hard to read after refactoring and prettier.